src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -258,51 +258,91 @@ import scala.runtime.Statics
-final class LazyList[+A] private(private[this] var lazyState: () => LazyList.State[A])
+final class LazyList[+A] private (lazyState: () => LazyList[A])
   extends AbstractSeq[A]
     with LinearSeq[A]
     with LinearSeqOps[A, LazyList, LazyList[A]]
     with IterableFactoryDefaults[A, LazyList]
     with Serializable {
   import LazyList._
 
-  @volatile private[this] var stateEvaluated: Boolean = false
-  @inline private def stateDefined: Boolean = stateEvaluated
-  private[this] var midEvaluation = false
-
-  private lazy val state: State[A] = {
-    // if it's already mid-evaluation, we're stuck in an infinite
-    // self-referential loop (also it's empty)
-    if (midEvaluation) {
-      throw new RuntimeException(
-        "LazyList evaluation depends on its own result (self-reference); see docs for more info"
-      )
+  private def this(head: A, tail: LazyList[A]) = {
+    this(LazyList.EmptyFun)
+    _head = head
+    _state = tail
+  }
+
+  // used to synchronize lazy state evaluation
+  // when `_head ne Uninitialized`
+  //   - `null` if this is an empty lazy list
+  //   - `head: A` otherwise (can be `null`, `_state == null` is used to test emptyness)
+  @volatile private[this] var _head: Any = Uninitialized
+
+  // when `_head eq Uninitialized`
+  //   - `lazySate: () => State`
+  //   - MidEvaluation while evaluating lazyState
+  // when `_head ne Uninitialized`
+  //   - `null` if this is an empty lazy list
+  //   - `tail: LazyList[A]` otherwise
+  private[this] var _state: AnyRef = lazyState
+
+  private def rawHead: Any = _head
+  private def rawState: AnyRef = _state
+
+  @inline private def stateDefined: Boolean = _head.asInstanceOf[AnyRef] ne Uninitialized
+
+  private def initState(): Unit = synchronized {
+    if (!stateDefined) {
+      // if it's already mid-evaluation, we're stuck in an infinite
+      // self-referential loop (also it's empty)
+      if (_state eq MidEvaluation)
+        throw new RuntimeException(
+          "LazyList evaluation depends on its own result (self-reference); see docs for more info")
+
+      val fun = _state.asInstanceOf[() => LazyList[A]]
+      _state = MidEvaluation
+      // restore `fun` in finally: consistent with previous 2.13 implementation which used a lazy val
+      val r = try fun.apply() finally _state = fun
+      r match {
+        case null =>
+          _state = null
+          _head = null
+        case l: LazyList[A @unchecked] =>
+          _state = l.rawState
+          _head = l.rawHead
+      }
     }
-    midEvaluation = true
-    val res = try lazyState() finally midEvaluation = false
-    // if we set it to `true` before evaluating, we may infinite loop
-    // if something expects `state` to already be evaluated
-    stateEvaluated = true
-    lazyState = null // allow GC
-    res
   }
 
+  private def state: LazyList[A] =
+    if (stateDefined) {
+      if (_state == null) LazyList.Empty
+      else this
+    } else {
+      initState()
+      state
```

Comment:
this should probably be explicitly marked `@tailrec` so someone doesn't accidentally make a change that allows it to overflow the stack